### PR TITLE
[Add] Adds accent selector

### PR DESCRIPTION
**Pull request by OctoRocket**
## About the PR

A selector for accents.

**Media**

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Added accent selector.

### CODE REVIEW

**Comment by deltanedas** [comment]
will this break cat ears / rat king crown

File: Content.Server/Traits/Assorted/AccentlessSystem.cs
Comment (ficcialfaint) [inline Content.Server/Traits/Assorted/AccentlessSystem.cs:23]: do ya really need that variable

File: Content.Server/Traits/Assorted/AccentlessSystem.cs
Comment (ficcialfaint) [inline Content.Server/Traits/Assorted/AccentlessSystem.cs:24]: maybe a better naming then just `t`?

File: Content.Server/Traits/Assorted/AccentlessSystem.cs
Comment (ficcialfaint) [inline Content.Server/Traits/Assorted/AccentlessSystem.cs:7]: 🫠 Don't leave it empty, make a simple description

File: Content.Server/Traits/Assorted/AccentlessSystem.cs
Comment (ficcialfaint) [inline Content.Server/Traits/Assorted/AccentlessSystem.cs:26]: There is no need for that, `RemoveComponent` has a built-in check

File: Content.Server/Traits/Assorted/AccentlessSystem.cs
Comment (ficcialfaint) [inline Content.Server/Traits/Assorted/AccentlessSystem.cs:27]: Just `RemComp` would be fine

**Review by ficcialfaint — CHANGES_REQUESTED** [review]

**Comment by OctoRocket** [comment]
No it would not break those

File: Content.Server/Traits/Assorted/AccentlessSystem.cs
Comment (OctoRocket) [reply Content.Server/Traits/Assorted/AccentlessSystem.cs]: I am not sure how I would be able to implement this

Comment (OctoRocket) [reply Content.Server/Traits/Assorted/AccentlessSystem.cs]: Nevermind.

**Comment by OctoRocket** [comment]
Everything is fixed!

File: Content.Server/Traits/Assorted/AccentlessSystem.cs
Comment (metalgearsloth) [inline Content.Server/Traits/Assorted/AccentlessSystem.cs:11]: This should be a datafield not hardcoded into the system.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Shared + access

File: Content.Server/Traits/Assorted/AccentlessComponent.cs
Comment (metalgearsloth) [inline Content.Server/Traits/Assorted/AccentlessComponent.cs:14]: Use compregistry and datafield plz.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Traits/Assorted/AccentlessComponent.cs
Comment (OctoRocket) [reply Content.Server/Traits/Assorted/AccentlessComponent.cs:14]: What do you mean by datafield?

Comment (OctoRocket) [reply Content.Server/Traits/Assorted/AccentlessComponent.cs:14]: I also don't know how I would use a component registry here.

File: RobustToolbox
Comment (ficcialfaint) [inline RobustToolbox:1]: Do not include submodules changes in your PR

**Review by ficcialfaint — CHANGES_REQUESTED** [review]

File: RobustToolbox
Comment (OctoRocket) [reply RobustToolbox:1]: I am having issues updating things apparently, I'll try to fix this if I can.

File: Content.Server/Traits/Assorted/AccentlessComponent.cs
Comment (metalgearsloth) [reply Content.Server/Traits/Assorted/AccentlessComponent.cs]: CompRegistry gives you the component types and is serializable and validates component existence.

Comment (OctoRocket) [reply Content.Server/Traits/Assorted/AccentlessComponent.cs]: It has been switched over.

**Comment by OctoRocket** [comment]
Everything should be fixed now, thanks for the suggestions!

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Still needs moving to shared + access attribute added to the component.

**Comment by OctoRocket** [comment]
Why does it need to be moved to shared?

**Comment by mirrorcult** [comment]
i think this is silly conceptually. why should you be able to opt out of random features like this

**Comment by OctoRocket** [comment]
I know that people find accents, specifically the dwarf accent, to be really hard to understand at times. It is also for the people where the accent doesn't fit their character, and so they don't want to have it.

**Comment by CloneTrooperSimp** [comment]
Personally, as a dwarf player, the accent is a tad annoying, but mostly just... really overwhelming. And yes, I have had several people be very confused by it, which does not help that. In my opinion, accents such as that shouldn't be a thing you automatically give to players, you should opt in, because unlike lizards, who all seem to follow the same species speech pattern, I doubt every dwarf, much like every human, would speak the same.

**Comment by OctoRocket** [comment]
The component and system have both been moved to shared!

**Comment by throge** [comment]
> Personally, as a dwarf player, the accent is a tad annoying, but mostly just... really overwhelming. And yes, I have had several people be very confused by it, which does not help that. In my opinion, accents such as that shouldn't be a thing you automatically give to players, you should opt in, because unlike lizards, who all seem to follow the same species speech pattern, I doubt every dwarf, much like every human, would speak the same.

Yeah, I agree here. The lizard accent exists because of the species itself. It makes sense for lizards to have a lisp due to their biology. Dwarves however basically just have a cooked Scottish accent, which, irl, is a cultural thing rather than species related. Dwarves, as a species, aren't limited to only a Scottish accent, and hence it makes sense for it to be something you can choose. This isn't even removing the accent in the first place, and quite frankly, I'd just prefer to do the accent myself if I felt it matched my character IC, and the benefit of this is I can pick and choose which words I want to apply the accent to.

tldr: Dwarf accent is pretty ehhh, makes no sense for it to be applied to every dwarf as a species (the accent is more of a cultural thing than biological). If you really want the accent, just don't activate the damn trait or just do the accent yourself.

**Comment by mirrorcult** [comment]
it is opt-in, you have to play dwarf to have it

**Comment by throge** [comment]
> it is opt-in, you have to play dwarf to have it

some just want to play dwarf to be short. works well for small human characters, plus it just overall looks funny.

**Comment by mirrorcult** [comment]
then add the ability to be short instead of 'you can be a dwarf but without actually being a dwarf'

**Comment by CloneTrooperSimp** [comment]
> then add the ability to be short instead of 'you can be a dwarf but without actually being a dwarf'

This was done before though, it was attempted rather, and was rejected by a majority of maintainers because of "fear of powergaming", since the short people could hide around things, which was a silly claim, people can hide regardless of their height in game.
Point is, an accentless trait feature is not gonna hurt anyone, it's not gonna be forced onto people, it's just a deal of you pick it or you don't. Most people enjoy the dwarf accent anyway, a few don't, that's it really, this feature is not gonna majorly change anything gameplay wise anyway, it's mostly for roleplay, which we honestly need more of.

**Comment by CrigCrag** [comment]
A way to change your height would probably be the better way to solve this. Disarms and shoves are essentially a mass contest so being a bit smaller is a disadvantage if anything. Being taller will need balancing, though, especially with Diona.

Being able to remove a characteristic of a race with a checkbox is silly.

**Comment by Emisse** [comment]
> then add the ability to be short instead of 'you can be a dwarf but without actually being a dwarf'

whats the difference

**Comment by CloneTrooperSimp** [comment]
> A way to change your height would probably be the better way to solve this. Disarms and shoves are essentially a mass contest so being a bit smaller is a disadvantage if anything. Being taller will need balancing, though, especially with Diona.
> 
> Being able to remove a characteristic of a species with a checkbox is silly.

Yeah, see, again, there was a PR for that before, namely this one #11488 and it was widely rejected by people, so I don't see how bringing it back would make things any better instead of just adding this simple option :/

**Comment by CrigCrag** [comment]
> > A way to change your height would probably be the better way to solve this. Disarms and shoves are essentially a mass contest so being a bit smaller is a disadvantage if anything. Being taller will need balancing, though, especially with Diona.
> > Being able to remove a characteristic of a species with a checkbox is silly.
> 
> Yeah, see, again, there was a PR for that before, namely this one #11488 and it was widely rejected by people, so I don't see how bringing it back would make things any better instead of just adding this simple option :/

Considering theirs an entire race based around size difference, I don't see how people could hate it that much anymore.

**Comment by EmoGarbage404** [comment]
The only difference between "accentless" and "short" traits is that one is a removal of a characteristic while the other is an addition. From a design perspective, a trait that adds something makes more sense than a trait that just removes something else.

Maybe its worth revisiting tall and short traits.

**Comment by Emisse** [comment]
pjb wanted displacement maps for short and tall

**Comment by OctoRocket** [comment]
I don't think people will like having the small trait because we already have Dwarves.

**Comment by OctoRocket** [comment]
I much rather have something that works now rather than wait for a PR that may or may not be made with changes that are very probably not going to be liked by the community.

**Comment by spectralSpices** [comment]
I believe that making an accentless trait for characters can only really add options for players in the future. If someone were to add a short/tall trait PR, I'd also support _that_, but I don't see what the problem with giving people character options would be. Does the game need it? Objectively, no. But does the game need a double-sided energy sword? Does the game _need_ expeditions? Does the game _need_ the HoP? 
It adds variety. It lets people make characters more unique. Or it's just a quality of life feature for people that may like playing as a species, but the accent is too distracting to them.

**Comment by CrigCrag** [comment]
methinks accents help identify people over radio (e.g. you see a lizard with no ID and then someone who isn't on the manifest talks with a lizard accent)

**Comment by spectralSpices** [comment]
> methinks accents help identify people over radio (e.g. you see a lizard with no ID and then someone who isn't on the manifest talks with a lizard accent)

Any species with an accent already has obvious features like a tail, roots, giant spider-legs, or being short.  I don't really understand what point you're trying to make here regarding the actual accentless trait.

**Comment by Matz05** [comment]
> > methinks accents help identify people over radio (e.g. you see a lizard with no ID and then someone who isn't on the manifest talks with a lizard accent)
> 
> Any species with an accent already has obvious features like a tail, roots, giant spider-legs, or being short. I don't really understand what point you're trying to make here regarding the actual accentless trait.

I believe they're saying that accentless removes (or, rather, replicates) the mechanical incentive to always play human (humans can verbally impersonate an accented species with a voice changer, but not the other way around) for optimal gameplay flexibility, especially as traitors where you'll often want to hide your identity.

**Comment by spectralSpices** [comment]
> > > methinks accents help identify people over radio (e.g. you see a lizard with no ID and then someone who isn't on the manifest talks with a lizard accent)
> > 
> > 
> > Any species with an accent already has obvious features like a tail, roots, giant spider-legs, or being short. I don't really understand what point you're trying to make here regarding the actual accentless trait.
> 
> I believe they're saying that accentless removes (or, rather, replicates) the mechanical incentive to always play human (humans can verbally impersonate an accented species with a voice changer, but not the other way around) for optimal gameplay flexibility, especially as traitors where you'll often want to hide your identity.

So wouldn't this be...better for the game, then? To make it so that players that may be other species can get more creative with their antagonist plays? Because the tone seemed rather critical.

**Comment by CloneTrooperSimp** [comment]
> > > > methinks accents help identify people over radio (e.g. you see a lizard with no ID and then someone who isn't on the manifest talks with a lizard accent)
> > > 
> > > 
> > > Any species with an accent already has obvious features like a tail, roots, giant spider-legs, or being short. I don't really understand what point you're trying to make here regarding the actual accentless trait.
> > 
> > 
> > I believe they're saying that accentless removes (or, rather, replicates) the mechanical incentive to always play human (humans can verbally impersonate an accented species with a voice changer, but not the other way around) for optimal gameplay flexibility, especially as traitors where you'll often want to hide your identity.
> 
> So wouldn't this be...better for the game, then? To make it so that players that may be other species can get more creative with their antagonist plays? Because the tone seemed rather critical.

Even then I doubt that'd happen anyway. The average player is going to be too explosives/trigger happy to even bother attempting to use something like that to their advantage, which makes this entire point a little odd at the end.

**Comment by CrigCrag** [comment]
if everyone went trigger happy kaboom mode we wouldn't see as many people use voice changers and such as we do now.

**Comment by OctoRocket** [comment]
I feel like voice masks should be able to allow people to impersonate accents and speech sounds anyways, but that is for a future PR.

**Comment by deltanedas** [comment]
nemesis changing voice mask to dwarf mode:

**Comment by OctoRocket** [comment]
I'll make the voice mask fix after this and another pr is merged, so the incentive to pick this trait for power-gaming reasons will be gone.

**Review by AJCM-git — CHANGES_REQUESTED** [review]
This should be properly implemented as part of character customization, not a trait. Accents are a generic system and so to properly implement this you need to add a drop down button that is dynamically populated by accents, this sounds daunting but it really isnt that hard, take a look at `_backpackButton/CBackpackButton` and how it is used. The implementation should be very similar to what the backpack selection does but enumerating all `ReplacementAccentPrototype` (maybe add a new "Selectable" or something datafield to the prototype to filter which ones belong in the character menu), this button would belong in a new "communication tab" or something but for now since we dont have languages, put it next to the species selection button. Do feel free to ask any question.

**Comment by spectralSpices** [comment]
So should Octo make a new PR?

**Comment by AJCM-git** [comment]
> So should Octo make a new PR?

They can just push to this one and update the title and description once they start updating it

**Comment by OctoRocket** [comment]
I was going to make an accent selector at first, but the ripple of errors that adding anything caused just made it look impossible.

**Comment by OctoRocket** [comment]
not what I wanted to do

**Comment by OctoRocket** [comment]
I seemingly can't change the branch name in the pr, so I'll have to make a new one it seems.

**Comment by AJCM-git** [comment]
The branch name is unchangeable but it really isnt that important compared to title/description

**Comment by OctoRocket** [comment]
Also, I can't enumerate over `ReplacementAccentPrototype`, because the majority of accents don't use `ReplacementAccentComponent`.

**Comment by OctoRocket** [comment]
I also don't think I can reopen this PR for whatever reason.

**Comment by AJCM-git** [comment]
> Also, I can't enumerate over `ReplacementAccentPrototype`, because the majority of accents don't use `ReplacementAccentComponent`.

You dont have to support every accent as a first implementation, the dwarf accent does have the component so you only need to support that one and "none" options for now, you can filter the prototypes by adding a new datafield to them

**Comment by spectralSpices** [comment]
> > Also, I can't enumerate over `ReplacementAccentPrototype`, because the majority of accents don't use `ReplacementAccentComponent`.
> 
> You dont have to support every accent as a first implementation, the dwarf accent does have the component so you only need to support that one and "none" options for now, you can filter the prototypes by adding a new datafield to them

Which means this could be ripe for expansion! Including accents that aren't even tied to any particular species or profession!

**Comment by OctoRocket** [comment]
That is what I'm planning on doing, moving the pirate accent into there too, but there doesn't seem to way to just get a list of all the accents. Some fun things I've found: There is a mobster ReplacementAccent, but it is only used for the rat king crown. The rats and the rat king instead use a system, `MobsterAccentSystem`, which is pretty dumb if you ask me. There is also the problem that I have no idea how to get a list of all the accents, while some of them are defined in yaml, like half of the mobster accent as well as the dwarf accent, the majority of them are just systems. I really don't want to hard code a list of accents anywhere, because that just seems like a terrible solution. If anyone has any ideas, I would love to hear them!

**Comment by OctoRocket** [comment]
This task is proving to be extraordinarily hard to complete.

**Comment by OctoRocket** [comment]
I'm sorry, but I'm not going to be working on this PR any longer. This PR was meant to be an accentless trait, and I worked on that for a while. When I ended up with the responsibility of making an accent selector instead, I accepted not knowing how difficult the task would be. I do not have the required skills to make the changes required for the accent selector, and so I am not going to be working on it.

**Comment by LankLTE** [comment]
I have to point out the irony in another accent trait (Archaic) being merged about 3 days after you were told you needed to add a selector if you wanted this in.

**Comment by Emisse** [comment]
i have no issue with the original version i think, i know mirror didnt like it but without tall/small i think its fine

**Comment by lzk228** [comment]
> If there comes a point where the accentless trait can be accepted, I would gladly reopen this PR.

so, accentless trait added
@OctoRocket

**Pull request closed without merging**
